### ISSUE TITLE
When set 'fold=all', how will the model to be trained?

### ISSUE THREAD

**Issue by WenshuaiZhao** [MIC-DKFZ/nnUNet#29]
Hi, Fabian:

You have said that the model by default to be trained in 5-fold cross-validation, I think it means if we set fold=number, then it will choose one kind of data_split, 1/5 data to be used as validation without being trained.  Then if we run all the 5 different split in the 5-fold cross-validation, we should save 5 models.

But I want to know if we set 'fold=all', how will it use the training data to get one model? I meant how it will split the data for validation and training.

Thank you!

**Comment by FabianIsensee** [comment]
Hi,
if you set fold='all' then all the training data will be used for training. Validation will run on the training data as well, but the numbers will be meaningless. 'all' is intended to be used if someone wants to compare their single-model (not ensemble) performance with nnU-Net on a separate test set.
Best,
Fabian

**Comment by WenshuaiZhao** [comment]
Fabian, thank you for your reply! 

Further, if you meant that the ensemble of the 5-fold cross validation models would be better than the single model trained by 'all' the data when tested by a new dataset? In other words, if we want to select the best model when applying on a new test dataset, should we use the ensemble one? or the single-model trained by 'fold=all'?

Additionally, I am a bit curious about the ensemble. How will the automatically selection script code select the best combination among the trained models, maybe 15 models totally? (2d, 3d, cascade)*(fold=0,1,2,3,4). OR is the combination just among one same model under different folds? and the selection for  2d, 3d, or cascade should be manual?

Sorry for the prolix expression. 

Best,
Wenshuai

**Comment by FabianIsensee** [comment]
Hi,
honestly, I don't have hard evidence suggesting that an ensemble of the cross-validation models performs better than a model trained on all data. My belief is that it is better. But to know for sure I would have to submit both variants to a test set - which would effectively be overfitting to the test set and thus inacceptable. Of course I could split off another validation set - but I don't have the time right now. If you want best performance, I think you should use the ensemble of the five folds.

After cross-validation, ensembling will be done with any combinations of two models: 2d &3d_lowres, 2d & 3d_fullres, 3d_lowres & 3d_fullres, 3d_fullres & 3d_cascade_fullres etc. These ensembles are done on the cross-validation data (you get a full set of predictions for the training set from those). Finally, the ensemble with the highest mean foreground dice (mean over foreground classes) will be selected as the one to be used for submission. It can also happen that a single model is best and in this case no model ensemble is used.
Whatever model(s) are selected, the test set predictions is done using the five folds of all selected models. So it's either 5 or 10 models.

All of this is automated as well but the code for this is not pretty.

Best,
Fabian

**Comment by WenshuaiZhao** [comment]
Thanks again for your detailed answer which helped me a lot! 

Best regards!
Wenshuai

**Comment by warmbasket** [comment]
>But to know for sure I would have to submit both variants to a test set - which would effectively be overfitting to the test set and thus inacceptable. 

How would you be overfitting to a test set? You didn't train on it?
